skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,203 @@
+        raise DescriptorException(
+            'Please ensure descriptors are a list of NumPy arrays.'
+        )
+
+    if isinstance(descriptors, list):
+        expected_shape = descriptors[0].shape
+        ranks = [len(e.shape) == len(expected_shape) for e in descriptors]
+        if (not all(ranks)):
+            raise DescriptorException(
+                'Please ensure all elements of your descriptor list '
+                'are of rank 2.'
+            )
+        dims = [e.shape[1] == descriptors[0].shape[1] for e in descriptors]
+        if not all(dims):
+            raise DescriptorException(
+                'Please ensure all descriptors are of the same dimensionality.'
+            )
+
+    if not isinstance(num_modes, int) or num_modes <= 0:
+        raise FisherVectorException(
+            'Please ensure num_modes is a positive integer.'
+        )
+
+    if 'covariance_type' in kwargs and kwargs['covariance_type'] != 'diag':
+        raise FisherVectorException('Covariane type must be "diag".')
+
+    if isinstance(descriptors, list):
+        descriptors = np.vstack(descriptors)
+
+    gmm = GaussianMixture(n_components=num_modes, covariance_type='diag')
+
+    gmm.fit(descriptors)
+
+    return gmm
+
+
+def compute_fisher_vector(descriptors, gmm):
+    """Function to compute the Fisher vector given some descriptors/vectors,
+    and an associated estimated GMM.
+
+    Parameters
+    -----------
+    descriptors : np.ndarray
+        NumPy array of the descriptors to use to compute the Fisher vector
+        representation for.
+    gmm : sklearn.mixture.GaussianMixture
+        An estimated GMM object, which contains the necessary parameters needed
+        to compute the Fisher vector.
+
+    Returns
+    -------
+    fisher_vector : np.ndarray
+        The computation Fisher vector, which is given by a concatenation of the
+        gradients of a GMM wrt its parameters (mixture weights, means, and
+        covariance matrices). For D-dimensional input descriptors/vectors, and
+        a K-mode GMM, the Fisher vector dimensionality will be 2KD + K. Thus,
+        its dimensionality is invariant to the number of descriptors/vectors.
+
+    References
+    -----------
+    .. [1] Perronnin, F. and Dance, C. Fisher kernels on Visual Vocabularies
+           for Image Categorization, IEEE Conference on Computer Vision and
+           Pattern Recognition, 2007
+
+    Examples
+    --------
+    >>> from skimage.feature import fisher_vector
+    >>> sift_for_images = [np.random.random((10, 128)) for _ in range(10)]
+    >>> num_modes = 16
+    >>> # Estimate 16-mode GMM with these synthetic SIFT vectors
+    >>> gmm = fisher_vector.learn_gmm(sift_for_images, num_modes)
+    >>> test_image_descriptors = np.random.random((25, 128))
+    >>> # Compute the Fisher vector
+    >>> fv = fisher_vector.compute_fisher_vector(test_image_descriptors, gmm)
+    """
+    from sklearn.mixture import GaussianMixture
```

Comment:
Since sklearn is an optional dependency (should be added as such in `requirements/optional.txt`), we should catch this importerror, print a sensible explanation, and then probably re-raise.

Reply:
@stefanv Done. I moved to optional dependencies and printed a sensible explanation when the import fails

Reply:
Done. I moved to optional dependencies and printed a sensible explanation when the import fails